tasks/el8/el8_postgresql.yml

```diff
@@ -0,0 +1,103 @@
+---
+- name: "Enable PostgreSQL DNF modules"
+  become: true
+  command:
+    cmd: dnf module -y enable postgresql
+    warn: false
+  tags:
+    - postgresql
+    - candlepin
+
+- name: "Install PostgreSQL packages"
+  become: true
+  package:
+    name:
+      - postgresql-server
+      - postgresql
+      - postgresql-jdbc
+      - python3-psycopg2
+    state: present
+  tags:
+    - postgresql
+    - candlepin
+
+- name: "Initialize PostgreSQL database"
+  become: true
+  command:
+    cmd: postgresql-setup --initdb --unit postgresql
+    creates: "/var/lib/pgsql/data/PG_VERSION"
+    warn: false
+  tags:
+    - postgresql
+    - candlepin
+
+- name: "Configure PostgreSQL to trust local connections (1 of 3)"
+  become: true
+  postgresql_pg_hba:
+    dest: /var/lib/pgsql/data/pg_hba.conf
+    contype: local
+    databases: all
+    users: all
+    method: trust
+    create: true
+    state: present
+  tags:
+    - postgresql
+    - candlepin
+
+- name: "Configure PostgreSQL to trust local connections (2 of 3)"
+  become: true
+  postgresql_pg_hba:
+    dest: /var/lib/pgsql/data/pg_hba.conf
+    contype: host
+    databases: all
+    users: all
+    address: ::1/128
+    method: trust
+    create: true
+    state: present
+  tags:
+    - postgresql
+    - candlepin
+
+- name: "Configure PostgreSQL to trust local connections (3 of 3)"
```

Comment:
This configuration does not allow connection from the host machine (for both el7/8). I get this one:
`[08001] Connection to 192.168.122.225:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.`

After I edit /var/lib/pgsql/data/postgresql.conf by adding `listen_addresses = '*'`, then I get this one:

`[28000] FATAL: no pg_hba.conf entry for host "192.168.122.1", user "candlepin", database "postgres", SSL off`

Then I had to edit /var/lib/pgsql/data/pg_hba.conf to add `host   all     all     0.0.0.0/0      trust`

Only then I can connect. These configurations were there previously, can we keep them?:

```
- name: Find host ip & netmask
  set_fact:
    host_address: "{{hostvars['dev']['ansible_default_ipv4']['gateway']}}"
    net_address_and_mask: "{{hostvars['dev']['ansible_default_ipv4']['gateway']}}/{{hostvars['dev']['ansible_default_ipv4']['netmask']}}"
  when: hostvars['dev'] is defined

- name: Set postgresql to trust connections from VM host
  lineinfile:
    dest: /var/lib/pgsql/12/data/pg_hba.conf
    line: "host    all         all         {{host_address}}/{{net_address_and_mask | ipaddr('prefix')}}           trust"
  when: hostvars['dev'] is defined
  tags:
    - candlepin
    - candlepin-root

- name: Set postgresql listen on all addresses
  lineinfile:
    dest: /var/lib/pgsql/12/data/postgresql.conf
    line: "listen_addresses = '*'"
  tags:
    - candlepin
    - candlepin-root
```


I know that this required to `pip install ansible netaddr` but instead of finding the host vm ip address, we could just approve all connections by setting `host   all     all     0.0.0.0/0      trust` as an alternative?

Reply:
Wait... are you connecting to the psql/mariadb servers in the guest from the host? That config makes sense from a Vagrant perspective, but I don't know if that's something we want shipping in a default ansible role, right?

I feel like such a thing starts running us into the issue we've hit before with containers and other stuff shared with QE and other departments where our needs conflict with theirs.

Reply:
Could this config not be made optional, and turned off by default? So that we can set that to being on through our Vagrantfile?

Reply:
Hrmm... I'm not sure how much value I see in it, but I've (temporarily?) added the following to get it to work:

```
# Make the databases accessible from remote machines
- name: "Check for the presence of PostgreSQL"
  become: true
  stat:
    path: "/var/lib/pgsql/data/pg_hba.conf"
  register: postgresql_conf_file
  tags:
    - postgresql
    - candlepin
    - debugging

- name: "Configure PostgreSQL to listen on all addresses"
  become: true
  lineinfile:
    regexp: "^#?\\s*listen_addresses\\s*="
    line: "listen_addresses = '*'"
    state: present
  when:
    - postgresql_conf_file.stat.exists
  tags:
    - postgresql
    - candlepin
    - debugging

- name: "Configure PostgreSQL to trust remote connections"
  become: true
  postgresql_pg_hba:
    dest: "{{postgresql_conf_file.stat.path}}"
    contype: host
    databases: all
    users: all
    address: "0.0.0.0/0"
    method: trust
    create: true
    state: present
  when:
    - postgresql_conf_file.stat.exists
  tags:
    - postgresql
    - candlepin
    - debugging

- name: "Restart the PostgreSQL service"
  become: true
  service:
    name: postgresql.service
    state: restarted
  when:
    - postgresql_conf_file.stat.exists
  tags:
    - postgresql
    - candlepin
    - debugging
```

Note that it involves hitting two config files to completely crack it open. I'm not a huge fan of this since it's so wide open, and the alternative provided is contingent on the host having a static IP or provisioning often enough to replace it (which itself has problems with the way the config task works). It's part of the debugging stuff, so the sheer insecurity of it is kind of expected, but it's a bit much for my taste given how automated this is.

Can you outline how often you use this? Do you do development from outside the VM and kinda treat it like a fat container?

Reply:
fact container sounds about right. I basically only ever run the deploy script and any gradle commands in the vm. Every other interaction is from my host (using intellij remote debugger, datagrip remote connection to the db, git directly on the host, etc.). That being said, my vagrant vms are usually long-lived (I might do vagrant up only once every ~5-6 months), so this is something I can live without.

I wonder what the rest of the team does @Januson @wolfdale @wottop : Is vagrant up preparing postgres for remote connections something you find useful, or you don't have a need for it ^?

Reply:
In fact, I might just re-use this ansible task list as a playbook to run after vagrant up is done (this will be enough to take away the pain of doing all this by hand)

Reply:
I wonder if there are some shenanigans we could do with git and the Vagrantfile where our repo includes a role that is .gitignore'd, which acts as a boilerplate for adding custom stuff we'd like to do with our images outside the reach of the candlepin role here so it wouldn't get overwritten by the galaxy bits, but would allow for some of the more obscure things we'd each like to do individually.

Reply:
I've added support for a custom role that will be automatically invoked during the usual provisioning process if it exists. This is done from the Vagrantfile's playbook, so it doesn't appear over here. That should suffice for all the custom stuff we want to do, and through symlinks and whatnot.

Reply:
the custom role works great, thanks